stan/math/prim/functor/closure_adapter.hpp

```diff
@@ -0,0 +1,190 @@
+#include <stan/math/prim/meta/is_stan_closure.hpp>
+#include <stan/math/prim/functor/apply.hpp>
+#include <ostream>
+
+namespace stan {
+namespace math {
+namespace internal {
+
+/**
+ * A closure that wraps a C++ lambda and captures values.
+ */
+template <bool Ref, typename F, typename... Ts>
+struct base_closure {
+  using captured_scalar_t__ = return_type_t<Ts...>;
+  using ValueOf__
+      = base_closure<false, F, decltype(eval(value_of(std::declval<Ts>())))...>;
+  using CopyOf__ = base_closure<false, F, Ts...>;
+  F f_;
+  std::tuple<capture_type_t<Ts, Ref>...> captures_;
+
+  explicit base_closure(const F& f, const Ts&... args)
+      : f_(f), captures_(args...) {}
+
+  template <typename... Args>
+  auto operator()(std::ostream* msgs, const Args&... args) const {
+    return apply([this, msgs, &args...](
+                     const auto&... s) { return f_(s..., args..., msgs); },
+                 captures_);
+  }
+};
+
+/**
+ * A closure that takes rng argument.
+ */
+template <bool Ref, typename F, typename... Ts>
+struct closure_rng {
+  using captured_scalar_t__ = double;
+  using ValueOf__ = closure_rng<false, F, Ts...>;
+  using CopyOf__ = closure_rng<false, F, Ts...>;
+  F f_;
+  std::tuple<capture_type_t<Ts, Ref>...> captures_;
+
+  explicit closure_rng(const F& f, const Ts&... args)
+      : f_(f), captures_(args...) {}
+
+  template <typename Rng, typename... Args>
+  auto operator()(Rng& rng, std::ostream* msgs, const Args&... args) const {
+    return apply([this, &rng, msgs, &args...](
+                     const auto&... s) { return f_(s..., args..., rng, msgs); },
+                 captures_);
+  }
+};
+
+/**
+ * A closure that can be called with `propto` template argument.
+ */
+template <bool Propto, bool Ref, typename F, typename... Ts>
+struct closure_lpdf {
+  using captured_scalar_t__ = return_type_t<Ts...>;
+  using ValueOf__ = closure_lpdf<Propto, false, F, Ts...>;
+  using CopyOf__ = closure_lpdf<Propto, false, F, Ts...>;
+  F f_;
+  std::tuple<capture_type_t<Ts, Ref>...> captures_;
+
+  explicit closure_lpdf(const F& f, const Ts&... args)
+      : f_(f), captures_(args...) {}
+
+  template <bool propto>
+  auto with_propto() {
+    return apply(
+        [this](const auto&... args) {
+          return closure_lpdf < Propto && propto, true, F,
+                 Ts... > (f_, args...);
+        },
+        captures_);
+  }
+
+  template <bool propto = false, typename... Args>
+  auto operator()(std::ostream* msgs, const Args&... args) const {
+    return apply(
+        [this, msgs, &args...](const auto&... s) {
+          return f_.template operator()<propto>(s..., args..., msgs);
+        },
+        captures_);
+  }
+};
+
+/**
+ * A closure that accesses logprob accumulator.
+ */
+template <bool Propto, bool Ref, typename F, typename... Ts>
+struct closure_lp {
+  using captured_scalar_t__ = return_type_t<Ts...>;
+  using ValueOf__ = closure_lp<Propto, true, F, Ts...>;
```

Comment:
Oops, these were supposed to be the same as in the base. It only works because they're never used.
No higher-order function works with `_rng` or `_lp` so they don't need either `ValueOf` or `CopyOf`. `reduce_sum` works with `_lpdf` but doesn't need `value_of()`.
I think these should just be removed.